optional_plugins/json_variants/avocado_json_variants/__init__.py

```diff
@@ -0,0 +1,133 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2018
+# Authors: Amador Pahim <apahim@redhat.com>
+
+import json
+import sys
+
+from six import iteritems
+
+from avocado.core import exit_codes
+from avocado.core import varianter
+from avocado.core.output import LOG_UI
+from avocado.core.plugin_interfaces import CLI
+from avocado.core.plugin_interfaces import Varianter
+
+
+class JsonVariantsCLI(CLI):
+
+    """
+    Serialized based Varianter options
+    """
+
+    name = 'json variants'
+    description = "JSON serialized based Varianter options for the 'run' " \
+                  "subcommand"
+
+    def configure(self, parser):
+
+        for name in ("run", "variants"):  # intentionally omitting "multiplex"
+            subparser = parser.subcommands.choices.get(name, None)
+            if subparser is None:
+                continue
+            sparser = subparser.add_argument_group('JSON serialized based '
+                                                   'varianter options')
+            sparser.add_argument('--json-variants-load', default=None,
+                                 help=('Load the Variants from a JSON '
+                                       'serialized file'))
+
+    def run(self, args):
+        pass
+
+
+class JsonVariants(Varianter):
+
+    """
+    Processes the serialized file into variants
+    """
+
+    name = 'json variants'
+    description = "JSON serialized based Varianter"
+
+    def initialize(self, args):
+        self.variants = None
+        load_variants = getattr(args, "json_variants_load", None)
+
+        if load_variants is None:
+            return
+        try:
+            with open(load_variants, 'r') as var_file:
+                self.variants = varianter.Varianter(state=json.load(var_file))
+        except IOError:
+            LOG_UI.error("JSON serialized file '%s' could not be found or "
+                         "is not readable", load_variants)
+            if args.subcommand == 'run':
+                sys.exit(exit_codes.AVOCADO_JOB_FAIL)
+            else:
+                sys.exit(exit_codes.AVOCADO_FAIL)
+
+    def __iter__(self):
+        if self.variants is None:
+            return
```

Comment:
The `self.variants` is defined during `initialize` and not on `__init__` (nor in class main), therefor this fails with `AttributeError: 'JsonVariants' object has no attribute 'variants'` and even if this is addressed I still believe it should not silently return as iterating on uninitialized JsonVariants makes no sense. Imagine someone wants to cache the number of variants so he simply calls `len(variants)` before `initialize` and blindly believes there are no variants. From my point of view this is dangerous and should raise RuntimeError, but if you have some special use-case that supports the no iterations, please share it here.

Reply:
The problem here is that we have in `avocado.core.Varianter` the `parse()` method which will populate the `self._no_variants` summing all the `len()` returns from all the plugins, so this plugin `__len__()` has to return something sum-able. As per your comment, you'd prefer to handle the exception on `parse()`, so I'm accommodating that change.

Reply:
Well, raising an exception here will make the dispatcher to handle it. We better return `None` and handle `None` in `avocado.core.Varianter.parse()`.